Freshen up CSS Highlight implementation

cdumez opened the pull request:
#### 4ed2d20fb3ad4f99b9d92288b3e82b5f26e23b4b
<pre>
Freshen up CSS Highlight implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=257520">https://bugs.webkit.org/show_bug.cgi?id=257520</a>

Reviewed by Wenson Hsieh and Darin Adler.

Freshen up CSS Highlight implementation:
- <a href="https://drafts.csswg.org/css-highlight-api-1/#highlight">https://drafts.csswg.org/css-highlight-api-1/#highlight</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/Highlight-setlike-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/Highlight-setlike.html: Added.
Import test coverage for Hightlight&apos;s SetLike behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/idlharness.window-expected.txt:
Rebaseline WPT test now that more checks are passing.

* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::repaintRange):
Take an AbstractRange instead of a SimpleRange for convenience, now that we support all
AbstractRange types.

(WebCore::Highlight::create):
(WebCore::Highlight::Highlight):
Update constructor to take in a FixedVector of AbstractRanges, to match the latest IDL.

(WebCore::Highlight::initializeSetLike):
Switch from StaticRange to AbstractRange to match the latest IDL.

(WebCore::Highlight::removeFromSetLike):
Do pointer comparison instead of using StaticRange&apos;s operator==(), to match JS&apos;s behavior.
This was causing Highlight-setlike.html to crash in debug.

(WebCore::Highlight::clearFromSetLike):
(WebCore::Highlight::repaint):
Use range getter now that HighlightRangeData&apos;s data members are private.

(WebCore::Highlight::addToSetLike):
The implementation was incorrect:
- It failed to move the range to the end if already present (since SetLike is an ordered set)
- It relies on StaticRange&apos;s operator==() instead of doing pointer comparison. In JS 2 staticRange
  objects are equal if they have the same pointer, no deep compare is happening. Our C++
  implementation needs to match this.

* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::HighlightRangeData::create):
(WebCore::HighlightRangeData::range const):
(WebCore::HighlightRangeData::startPosition const):
(WebCore::HighlightRangeData::endPosition const):
(WebCore::HighlightRangeData::HighlightRangeData):
- Make it a class instead of a struct, with a private constructor and private
  data members, since this subclasses RefCounted. Refcounted classes should
  never have a public constructor and it is too error-prone.
- Add public getters / setters now that the data members are private.
- Store an AbstractRange instead of a StaticRange to match the latest IDL.

(WebCore::Highlight::type const):
(WebCore::Highlight::setType):
(WebCore::Highlight::priority const):
(WebCore::Highlight::setPriority):
Add basic support for `type` and `priority` IDL attributes.

* Source/WebCore/Modules/highlight/Highlight.idl:
Synchronize with the specification:
- <a href="https://drafts.csswg.org/css-highlight-api-1/#highlight">https://drafts.csswg.org/css-highlight-api-1/#highlight</a>
In particular, we now use AbstractRange instead of StaticRange
and the `type` &amp; `priority` attributes were added. Also, the
constructor can now take in several ranges, not just one.

* Source/WebCore/Modules/highlight/HighlightRegister.cpp:
(WebCore::HighlightRegister::addAnnotationHighlightWithRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateHighlightPositions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/rendering/HighlightData.cpp:
(WebCore::HighlightData::setRenderRange):
* Source/WebCore/rendering/HighlightData.h:

Canonical link: <a href="https://commits.webkit.org/264739@main">https://commits.webkit.org/264739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6c3a48ff05768f157ee227205f100cf3dec71db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8485 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8997 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10152 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8531 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10768 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8726 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11399 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9673 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10307 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6971 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15314 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11267 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8384 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7670 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7675 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2058 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11881 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8129 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -30,69 +30,77 @@
-        set.add<IDLInterface<StaticRange>>(rangeData->range);
+    for (auto& range : initialRanges) {
+        repaintRange(range.get());
+        m_rangesData.append(HighlightRangeData::create(Ref { range.get() }));
```
Nit - reserveInitialCapacity + uncheckedAppend here?

darinadler replied:
I considered making this comment, but then I realized that there can be further appending to this vector later, and it might not be optimal to right-size it, so I did not suggest it. Not saying that you’re incorrect, just that it’s not totally obvious.

whsieh replied:
Ah, I see — that makes sense. I suppose my thought process was that, since the number of elements we're appending in this ctor is known ahead of time, there wouldn't be any disadvantage to resizing eagerly here. Though I suppose, if any highlights are added later down the road, then any extra capacity might help us avoid a resize then..

whsieh commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -30,69 +30,77 @@
+        m_rangesData.append(HighlightRangeData::create(range));
+        return true;
+    }
+    // Move to last since SetLike is an ordered set.
```
Should we consider making it a ListHashSet at some point? (ordered + fast arbitrary lookup)

cdumez replied:
Yes, a ListHashSet is a good idea.

cdumez replied:
That said, it would require some work because we want to look up AbstractRanges, not HighlightRangeData pointers.

whsieh replied:
Yep! Definitely a "separate-patch" kind of change.

rr-codes commented on Source/WebCore/Modules/highlight/Highlight.h:
```diff
@@ -32,42 +32,60 @@
 class Highlight : public RefCounted<Highlight> {
 public:
-    WEBCORE_EXPORT static Ref<Highlight> create(StaticRange&);
+    WEBCORE_EXPORT static Ref<Highlight> create(FixedVector<std::reference_wrapper<WebCore::AbstractRange>>&&);
```
what's the difference between a FixedVector and an std::array?

cdumez replied:
This is the type the bindings are passing us here so I cannot use a std::array. Unlike std::array, the length of a FixedVector is set at runtime, not compile time. std::array actually has the length as part of its type. FixedVector doesn't but the size doesn't change once constructed.

rr-codes replied:
ah, I see! thanks for the explanation!

cdumez replied:
In other words, we couldn't use a std::array here since we don't know at compile time how many ranges the JS will pass to the constructor.

darinadler commented on Source/WebCore/Modules/highlight/Highlight.h:
```diff
@@ -32,42 +32,60 @@
-    std::optional<Position> endPosition;
+
+    Ref<AbstractRange> m_range;
+    std::optional<Position> m_startPosition;
```
Since `Position` has a null value, surprised we need to use `std:.optional`. Also wondering why we don’t use the simpler `BoundaryPoint` type.

cdumez replied:
Will use Position for now as this patch is already big and @jesxilin is working on this. I'll let her follow up if needed to adopt `BoundaryPoint`.

whsieh reviewed (approved):
LGTM, if we can address the layout test failure.

cdumez committed:
Freshen up CSS Highlight implementation
https://bugs.webkit.org/show_bug.cgi?id=257520

Reviewed by Wenson Hsieh and Darin Adler.

Freshen up CSS Highlight implementation:
- https://drafts.csswg.org/css-highlight-api-1/#highlight

* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/Highlight-setlike-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/Highlight-setlike.html: Added.
Import test coverage for Hightlight's SetLike behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/idlharness.window-expected.txt:
Rebaseline WPT test now that more checks are passing.

* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::repaintRange):
Take an AbstractRange instead of a SimpleRange for convenience, now that we support all
AbstractRange types.

(WebCore::Highlight::create):
(WebCore::Highlight::Highlight):
Update constructor to take in a FixedVector of AbstractRanges, to match the latest IDL.

(WebCore::Highlight::initializeSetLike):
Switch from StaticRange to AbstractRange to match the latest IDL.

(WebCore::Highlight::removeFromSetLike):
Do pointer comparison instead of using StaticRange's operator==(), to match JS's behavior.
This was causing Highlight-setlike.html to crash in debug.

(WebCore::Highlight::clearFromSetLike):
(WebCore::Highlight::repaint):
Use range getter now that HighlightRangeData's data members are private.

(WebCore::Highlight::addToSetLike):
The implementation was incorrect:
- It failed to move the range to the end if already present (since SetLike is an ordered set)
- It relies on StaticRange's operator==() instead of doing pointer comparison. In JS 2 staticRange
  objects are equal if they have the same pointer, no deep compare is happening. Our C++
  implementation needs to match this.

* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::HighlightRangeData::create):
(WebCore::HighlightRangeData::range const):
(WebCore::HighlightRangeData::startPosition const):
(WebCore::HighlightRangeData::endPosition const):
(WebCore::HighlightRangeData::HighlightRangeData):
- Make it a class instead of a struct, with a private constructor and private
  data members, since this subclasses RefCounted. Refcounted classes should
  never have a public constructor and it is too error-prone.
- Add public getters / setters now that the data members are private.
- Store an AbstractRange instead of a StaticRange to match the latest IDL.

(WebCore::Highlight::type const):
(WebCore::Highlight::setType):
(WebCore::Highlight::priority const):
(WebCore::Highlight::setPriority):
Add basic support for `type` and `priority` IDL attributes.

* Source/WebCore/Modules/highlight/Highlight.idl:
Synchronize with the specification:
- https://drafts.csswg.org/css-highlight-api-1/#highlight
In particular, we now use AbstractRange instead of StaticRange
and the `type` & `priority` attributes were added. Also, the
constructor can now take in several ranges, not just one.

* Source/WebCore/Modules/highlight/HighlightRegister.cpp:
(WebCore::HighlightRegister::addAnnotationHighlightWithRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateHighlightPositions):
* Source/WebCore/dom/Document.h:
* Source/WebCore/rendering/HighlightData.cpp:
(WebCore::HighlightData::setRenderRange):
* Source/WebCore/rendering/HighlightData.h:

Canonical link: https://commits.webkit.org/264739@main

rkirsling commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -74,7 +74,7 @@ void HighlightRegister::addAnnotationHighlightWithRange(Ref<StaticRange>&& value
         m_map.get(annotationHighlightKey())->addToSetLike(value);
     else
-        setFromMapLike(annotationHighlightKey(), Highlight::create(WTFMove(value)));
+        setFromMapLike(annotationHighlightKey(), Highlight::create({ WTFMove(value) }));
```
This doesn't build on PS, but it's taking me a moment to figure out why, as it's quite hard to understand what the braces are precisely meant to refer to. Is the intention that everything but the initializer_list happens implicitly?

cdumez replied:
Yes, the idea is that it would call the following constructor:
```
FixedVector(std::initializer_list<T> initializerList);
```
May need to use `FixedVector({ ... })` for PS or even `FixedVector<std::reference_wrapper<WebCore::AbstractRange>>({ ... })`.

rkirsling replied:
Hmm, I guess that may not be the source of the issue at all. The confusing part to me is how Ref becomes std::reference_wrapper, and indeed, it seems like that's what our compiler is stuck on. Guess I'll check locally whether `{ std::ref(value.get()) }` works.
